File: Samples/OCR/cpp/WordOverlay.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "WordOverlay.h"

using namespace SDKTemplate;

WordOverlay::WordOverlay(OcrWord^ ocrWord) :
    word(ocrWord),
    isPropertyChangedObserved(false)
{
    UpdateProps(word->BoundingRect);
}

void WordOverlay::Transform(ScaleTransform^ transform)
{
    // Scale word box bounding rect and update properties.
    UpdateProps(transform->TransformBounds(word->BoundingRect));
}

void WordOverlay::UpdateProps(Rect boundingBox)
{
    wordPosition = *(ref new Thickness(boundingBox.Left, boundingBox.Top, 0, 0));
    wordWidth = boundingBox.Width;
    wordHeight = boundingBox.Height;

    OnPropertyChanged("WordPosition");
    OnPropertyChanged("WordWidth");
    OnPropertyChanged("WordHeight");
}

/// <summary>
/// Notifies listeners that a property value has changed. 
/// </summary>
/// <param name="propertyName">Name of the property used to notify listeners.</param> 
void WordOverlay::OnPropertyChanged(Platform::String^ propertyName)
{
    if (isPropertyChangedObserved)
    {
        PropertyChanged(this, ref new PropertyChangedEventArgs(propertyName));
    }
}

Binding^ WordOverlay::CreateWordPositionBinding()
{
    Binding^ positionBinding = ref new Binding();
    positionBinding->Path = ref new PropertyPath("WordPosition");
    positionBinding->Source = this;

    return positionBinding;
}

Binding^ WordOverlay::CreateWordWidthBinding()
{
    Binding^ widthBinding = ref new Binding();
    widthBinding->Path = ref new PropertyPath("WordWidth");
    widthBinding->Source = this;

    return widthBinding;
}

Binding^ WordOverlay::CreateWordHeightBinding()
{
    Binding^ heightBinding = ref new Binding();
    heightBinding->Path = ref new PropertyPath("WordHeight");
    heightBinding->Source = this;

    return heightBinding;
}

File: Samples/OCR/cs/WordOverlay.cs
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

using System.ComponentModel;
using Windows.Foundation;
using Windows.Media.Ocr;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Data;
using Windows.UI.Xaml.Media;

namespace SDKTemplate
{
    /// <summary>
    /// Wraps recognized word and provides word overlay position and size based on current scale.
    /// </summary>
    class WordOverlay : INotifyPropertyChanged
    {
        private OcrWord word;

        /// <summary>
        /// Left and Right properties of Thickess define word box position.
        /// </summary>
        public Thickness WordPosition { get; private set; }

        /// <summary>
        /// Scaled word box width.
        /// </summary>
        public double WordWidth { get; private set; }

        /// <summary>
        /// Scaled word box height.
        /// </summary>
        public double WordHeight { get; private set; }

        public event PropertyChangedEventHandler PropertyChanged;

        public WordOverlay(OcrWord ocrWord)
        {
            word = ocrWord;

            UpdateProps(word.BoundingRect);
        }

        public void Transform(ScaleTransform scale)
        {
            // Scale word box bounding rect and update properties.
            UpdateProps(scale.TransformBounds(word.BoundingRect));
        }

        public Binding CreateWordPositionBinding()
        {
            Binding positionBinding = new Binding()
            {
                Path = new PropertyPath("WordPosition"),
                Source = this
            };

            return positionBinding;
        }

        public Binding CreateWordWidthBinding()
        {
            Binding widthBinding = new Binding()
            {
                Path = new PropertyPath("WordWidth"),
                Source = this
            };

            return widthBinding;
        }

        public Binding CreateWordHeightBinding()
        {
            Binding heightBinding = new Binding()
            {
                Path = new PropertyPath("WordHeight"),
                Source = this
            };

            return heightBinding;
        }

        private void UpdateProps(Rect wordBoundingBox)
        {
            WordPosition = new Thickness(wordBoundingBox.Left, wordBoundingBox.Top, 0, 0);
            WordWidth = wordBoundingBox.Width;
            WordHeight = wordBoundingBox.Height;

            OnPropertyChanged("WordPosition");
            OnPropertyChanged("WordWidth");
            OnPropertyChanged("WordHeight");
        }

        protected void OnPropertyChanged(string PropertyName)
        {
            if (PropertyChanged != null)
            {
                PropertyChanged(this, new PropertyChangedEventArgs(PropertyName));
            }
        }
    }
}
